### ISSUE TITLE
Question about Dapper and nonprimitives

### ISSUE THREAD

**Issue by robbert229** [StackExchange/dapper-dot-net#372]
So how does Dapper work with nonprimitives?

**Comment by mrinalkamboj** [comment]
What does this mean, non primitive means reference types, it works as expected, `Query<T>` will directly provide `IEnumerable<T>`. Not sure if I really understood everything in your question

**Comment by robbert229** [comment]
Let me rephrase my question. Say we are querying for an author object. Can the author object contain a reference to an address, or other objects? If so what does it look like code wise. 

Ex. 
    class Author {
        Address Address { get; set; }
        ....
    }

**Comment by mrinalkamboj** [comment]
Yes, now its much more clear, and the response is:

- For one to one complex mapping as in your example, its out of box, though it requires calling an overload of Query, which needs a `Func`, to map various entities fetched as part of select, also it needs setting `splitOn` to tell the Dapper, where one entity ends and other begins in the complete result, first Split is always at first column, remaining need to be specified. Check the following example:

`const string query = @"select p.BusinessEntityId
                            ,p.PersonType
                            ,p.NameStyle
                            ,p.Title
                            ,p.FirstName
                            ,p.MiddleName
                            ,p.LastName
                            ,p.Suffix
                            ,p.EmailPromotion
                            ,pa.PasswordHash
                            ,pa.PasswordSalt
                            ,pc.CreditCardID
                            ,pc.ModifiedDate
							,pp.PhoneNumber
							,pp.PhoneNumberTypeID                            
                            from [Person].person p join [Person].password pa on p.BusinessEntityID = pa.BusinessEntityID
                                 join [Sales].personCreditCard pc on p.BusinessEntityID = pc.BusinessEntityID
								 join [Person].PersonPhone pp on p.BusinessEntityID = pp.BusinessEntityID";

            var personPasswords = _connection.Query<PersonDetail, PersonPassword, MyEntities.PersonCreditCard, MyEntities.PersonPhone, PersonDetail>(query,
                (pd,pp,pc,pph) =>
                {
                    pd.Password = pp;
                    pd.CreditCard = pc;
                    pd.Phone = pph;
                    return pd;
                }, splitOn: "PasswordHash,CreditCardID,PhoneNumber").ToList();`

**Comment by mrinalkamboj** [comment]
In case you want one to many mapping like `List<Address>`, then its not out of box, you either need a Linq, after calling `QueryMultiple`, or you may use following generic extension method of the Dapper `GridReader`:

` public static IEnumerable<TFirst> Map<TFirst, TSecond, TKey>
        (
        this SqlMapper.GridReader reader,
        Func<TFirst, TKey> firstKey,
        Func<TSecond, TKey> secondKey,
        Action<TFirst, IEnumerable<TSecond>> addChildren
        )
        {
            var first = reader.Read<TFirst>().ToList();

            var childMap = reader.Read<TSecond>()
                                 .GroupBy(secondKey)
                                 .Where(grp =>
                                 {
                                     if (
                                         Nullable.GetUnderlyingType(typeof(TKey)) != null 
                                         && grp.Key == null
                                         )
                                         return false;
                                     return true;
                                 })
                                 .ToDictionary(g => g.Key, g => g.ToList());

          
            foreach (var item in first)
            {
                List<TSecond> children;

                addChildren(item, childMap.TryGetValue(firstKey(item), out children) ? children : new List<TSecond>());
            }

            return first;
        }`

Usage of the same is:

`gridReader.Read<TFirst>().Map`
and Result would be `IEnumerable<TFirst>` with one to many mapping done

**Comment by NickCraver** [comment]
@robbert229  to be more concise, you want what we call a "multi-map query", like this:

```C#
var authors = conn.Query<Author, Address, Author>(@"
    Select a.*, add.* From Authors a Join Addresses add On a.AddressId = add.Id", 
    (a, add) => { a.Address = add; return a; });
```
The above assumes Addresses starts with an `Id` column. If this isn't true then you just set the optional `splitOn` param on this method, e.g. `splitOn:"AuthorId,AddressId"` if that's your Id naming scheme.

**Comment by robbert229** [comment]
Thanks. This is exactly what I was looking for.
